server/src/test/java/org/elasticsearch/index/replication/RetentionLeasesReplicationTests.java

```diff
@@ -102,50 +104,56 @@ public void testSyncRetentionLeasesWithPrimaryPromotion() throws Exception {
         Settings settings = Settings.builder().put(IndexSettings.INDEX_SOFT_DELETES_SETTING.getKey(), true).build();
         int numberOfReplicas = between(2, 4);
         IndexMetaData indexMetaData = buildIndexMetaData(numberOfReplicas, settings, indexMapping);
+        final AtomicReference<Consumer<RetentionLeaseSyncAction.Request>> onRequest = new AtomicReference<>();
         try (ReplicationGroup group = new ReplicationGroup(indexMetaData) {
             @Override
-            protected void syncRetentionLeases(ShardId shardId, RetentionLeases leases, ActionListener<ReplicationResponse> listener) {
-                listener.onResponse(new SyncRetentionLeasesResponse(new RetentionLeaseSyncAction.Request(shardId, leases)));
+            protected void syncRetentionLeases(ShardId shardId, RetentionLeases leases, ActionListener<Void> listener) {
+                onRequest.get().accept(new RetentionLeaseSyncAction.Request(shardId, leases));
+                listener.onResponse(null);
             }
         }) {
             group.startAll();
             int numLeases = between(1, 100);
-            IndexShard newPrimary = randomFrom(group.getReplicas());
-            RetentionLeases latestRetentionLeasesOnNewPrimary = RetentionLeases.EMPTY;
-            for (int i = 0; i < numLeases; i++) {
-                PlainActionFuture<ReplicationResponse> addLeaseFuture = new PlainActionFuture<>();
-                group.addRetentionLease(Integer.toString(i), randomNonNegativeLong(), "test-" + i, addLeaseFuture);
-                RetentionLeaseSyncAction.Request request = ((SyncRetentionLeasesResponse) addLeaseFuture.actionGet()).syncRequest;
+            final IndexShard newPrimary = randomFrom(group.getReplicas());
+            final AtomicReference<RetentionLeases> latestRetentionLeasesOnNewPrimary = new AtomicReference<>(RetentionLeases.EMPTY);
```

Comment:
This test becomes a bit more tangled since we change to a Void listener.

Reply:
It does. I nearly backed out the whole idea because of this test. OTOH it has less casting after this change.